fuel-vm/src/interpreter/debug.rs

```diff
@@ -63,140 +57,192 @@ where
+            .finalize()
+            .into_checked(height, &consensus_params)
+            .expect("failed to generate checked tx")
+            .into_ready(
+                gas_price,
+                consensus_params.gas_costs(),
+                consensus_params.fee_params(),
+            )
+            .unwrap();
+
+        let suite = vec![
+            (
+                Breakpoint::script(0),
+                vec![(0x10, 0), (0x11, 0), (0x12, 0), (0x13, 0), (0x14, 0)],
+            ),
+            (
+                Breakpoint::script(2),
+                vec![(0x10, 8), (0x11, 16), (0x12, 0), (0x13, 0), (0x14, 0)],
+            ),
+            (
+                Breakpoint::script(3),
+                vec![(0x10, 8), (0x11, 16), (0x12, 32), (0x13, 0), (0x14, 0)],
+            ),
+            (
+                Breakpoint::script(5),
+                vec![(0x10, 8), (0x11, 16), (0x12, 32), (0x13, 64), (0x14, 128)],
+            ),
+        ];
+
+        suite.iter().for_each(|(b, _)| vm.set_breakpoint(*b));
+
+        let state = vm
+            .transact(tx)
+            .map(ProgramState::from)
+            .expect("Failed to execute script!");
+
+        suite
+            .into_iter()
+            .fold(state, |state, (breakpoint, registers)| {
+                let debug = state.debug_ref().expect("Expected breakpoint");
+                let b = debug
+                    .breakpoint()
+                    .expect("State without expected breakpoint");
+
+                assert_eq!(&breakpoint, b);
+                registers.into_iter().for_each(|(r, w)| {
+                    assert_eq!(w, vm.registers()[r]);
+                });
+
+                vm.resume().expect("Failed to resume")
+            });
+    }
+
+    #[test]
+    fn single_stepping() {
+        use fuel_asm::op;
+        use fuel_tx::ConsensusParameters;
+
+        let mut vm = Interpreter::<_, _, _>::with_memory_storage();
+
+        let gas_limit = 1_000_000;
+        let height = Default::default();
+        let gas_price = 0;
+
+        // Repeats the middle two instructions five times
+        let script = [
+            op::addi(0x10, RegId::ZERO, 5),
+            op::addi(0x11, 0x11, 1),
+            op::jnei(0x10, 0x11, 1),
+            op::ret(0x10),
+        ]
         .into_iter()
-        .fold(state, |state, (breakpoint, registers)| {
-            let debug = state.debug_ref().expect("Expected breakpoint");
+        .collect();
+
+        let consensus_params = ConsensusParameters::standard();
+
+        let tx = TransactionBuilder::script(script, vec![])
+            .script_gas_limit(gas_limit)
+            .add_random_fee_input()
+            .finalize()
+            .into_checked(height, &consensus_params)
+            .expect("failed to generate checked tx")
+            .into_ready(
+                gas_price,
+                consensus_params.gas_costs(),
+                consensus_params.fee_params(),
+            )
+            .unwrap();
```

Comment:
It seems we repeat this part several times, maybe makes sense to move it into a function